finetuning-llama-7b-on-l4/README.md

```diff
@@ -0,0 +1,381 @@
+
+
+Run the following commands to set the env variables and make sure to replace `<my-project-id>`:
+
+```bash
+gcloud config set project <my-project-id>
+export PROJECT_ID=$(gcloud config get project)
+export REGION=us-central1
+export BUCKET_NAME=${PROJECT_ID}-llama-l4
+export SERVICE_ACCOUNT="l4-demo@${PROJECT_ID}.iam.gserviceaccount.com"
+export IMAGE=gcr.io/$PROJECT_ID/llama2-finetune
+```
+
+Note: You might have to rerun the export commands if for some reason you reset your shell and the variables are no longer set. This can happen for example when your Cloud Shell disconnects.
+
+Create the GKE cluster by running:
+```bash
+gcloud container clusters create l4-demo --location ${REGION} \
+  --workload-pool ${PROJECT_ID}.svc.id.goog \
+  --enable-image-streaming --enable-shielded-nodes \
+  --shielded-secure-boot --shielded-integrity-monitoring \
+  --enable-ip-alias \
+  --node-locations=$REGION-a \
+  --workload-pool=${PROJECT_ID}.svc.id.goog \
+  --addons GcsFuseCsiDriver
+```
+
+(Optional) In environments where external IP addresses are not allowed you can add the following arguments to the create GKU cluster command:
+```bash
+  --no-enable-master-authorized-networks --enable-ip-alias \
+    --enable-private-nodes  --master-ipv4-cidr 172.16.0.32/28
+```
+
+Let’s create a nodepool for our finetuning which will use 8 L4 GPUs per VM:
+
+Create the `g2-standard-96` nodepool by running:
+```bash
+gcloud container node-pools create g2-standard-96 --cluster l4-demo \
+  --accelerator type=nvidia-l4,count=8,gpu-driver-version=latest \
+  --machine-type g2-standard-96 \
+  --ephemeral-storage-local-ssd=count=8 \
+  --enable-autoscaling --enable-image-streaming \
+  --num-nodes=0 --min-nodes=0 --max-nodes=3 \
+  --shielded-secure-boot \
+  --shielded-integrity-monitoring \
+  --node-locations $REGION-a,$REGION-b --region $REGION
+```
+
+
+NOTE: The `--node-locations` flag might have to be adjusted based on which region you choose. Please take a look at which zones the [L4 GPUs are available](https://cloud.google.com/compute/docs/gpus/gpu-regions-zones) if you change the region to something other than `us-central1`.
+
+The nodepool has been created and is scaled down to 0 nodes. So you are not paying for any GPUs until you start launching K8s Pods that request GPUs.
+
+## Run a K8s job to finetune Llama 2 7B
+
+Finetuning requires a base model and a dataset. For this post, the [dell-research-harvard/AmericanStories](https://huggingface.co/datasets/dell-research-harvard/AmericanStories) dataset will be used to finetune the [Llama 2 7B](https://huggingface.co/meta-llama/Llama-2-7b-hf) base model. GCS will be used for storing the base model. GKE with GCSFuse is used to transparently save the finetuned model to GCS. This provides a cost efficient way to store and serve the model and only pay for the storage used by the model.
```

Comment:
I store a fully merged model because not all other libraries support LoRA. I think it often over complicates things. I added a sentence to explain this when we introduce finetune.py

Reply:
I had a lot of issues when using HuggingFace models where I only had the LoRA layer that referred to a base model.

I wrote this in the guide:
The fine-tuned model initially are separate LoRA weights. In the `finetune.py` script, the base model and LoRA weights are merged
so the fine-tuned model can be used as a standalone model. This does utilize more storage than needed,
but in return you get better compatibility with different libraries for serving.